lua/lib/lfospec.lua

```diff
@@ -0,0 +1,449 @@
+  params:add{
+    type = 'control',
+    id = 'lfo_'..bound..'_'..i,
+    name = 'lfo '..bound,
+    controlspec = controlspec.new(
+      lfo.min,
+      lfo.max,
+      lfo.controlspec.warp,
+      lfo.controlspec.step,
+      bound == 'min' and lfo.min or lfo.max,
+      lfo.controlspec.units,
+      lfo.controlspec.quantum,
+      lfo.controlspec.wrap
+    ),
+    formatter = lfo.controlspec.formatter
+  }
+  params:set_action('lfo_'..bound..'_'..i, function(x)
+    lfo:set(bound,x)
+  end)
+end
+
+local function lfo_bang(i)
+  local function lb(prm)
+    params:lookup_param("lfo_"..prm.."_"..i):bang()
+  end
+  lb('depth')
+  lb('min')
+  lb('max')
+  lb('baseline')
+  lb('offset')
+  lb('mode')
+  lb('clocked')
+  lb('free')
+  lb('shape')
+  lb('reset')
+  lb('reset_target')
+end
+
+-- / PARAMETERS UI
+
+-- SCRIPTING /
+
+local function process_lfo(id)
+  local _lfo = id
+  local phase
+
+  _lfo.phase_counter = _lfo.phase_counter + (1/_lfo.ppqn)
+  if _lfo.mode == "clocked" then
+    phase = _lfo.phase_counter / _lfo.period
+  else
+    phase = _lfo.phase_counter * clock.get_beat_sec() / (1/_lfo.period)
+  end
+  phase = phase % 1
+
+  if _lfo.enabled == 1 then
+
+    local current_val = (math.sin(2*math.pi*phase) + 1)/2
+    if _lfo.shape == "saw" then
+      current_val = phase < 0.5 and phase/0.5 or 1-(phase-0.5)/(0.5)
+    end
+
+    local min = _lfo.min
+    local max = _lfo.max
+
+    if min > max then
+      local old_min = min
+      local old_max = max
+      min = old_max
+      max = old_min
+    end
+
+    if min == -inf or max == -inf then
+      min = min == -inf and -2^1023.999999999 or min
+      max = max == inf and 2^1023.999999999 or max
+    end
+
+    local percentage = math.abs(min-max) * _lfo.depth
+    if _lfo.shape ~= 'random' then
+      _lfo.raw = current_val
+    end
+    current_val = current_val + _lfo.offset
+    local value = util.linlin(0,1,min,min + percentage,current_val)
+
+    if _lfo.depth > 0 then
+      local mid;
+      local scaled_min;
+      local scaled_max;
+      local raw_value;
+
+      if _lfo.baseline == 'min' then
+	scaled_min = min
+	scaled_max = min + percentage
+	mid = util.linlin(min,max,scaled_min,scaled_max,(min+max)/2)
+      elseif _lfo.baseline == 'center' then
+	mid = (min+max)/2
+	local centroid_mid = math.abs(min-max) * ((_lfo.depth)/2)
+	scaled_min = util.clamp(mid - centroid_mid,min,max)
+	scaled_max = util.clamp(mid + centroid_mid,min,max)
+	value = util.linlin(0,1,scaled_min,scaled_max,current_val)
+      elseif _lfo.baseline == 'max' then
+	mid = (min+max)/2
+	value = max - value
+	scaled_min = max * (1-(_lfo.depth))
+	scaled_max = max
+	mid = math.abs(util.linlin(min,max,scaled_min,scaled_max,mid))
+	value = util.linlin(0,1,scaled_max,scaled_min,current_val)
+      end
+
+      raw_value = util.linlin(min,max,0,1,value)
+
+      if _lfo.shape == "sine" then
+	value = util.clamp(value,min,max)
+	_lfo.scaled = value
+      elseif _lfo.shape == "saw" then
```

Comment:
- i would refactor this whole conditional change into spearate functions per mode, and one function handle that is assigned when the mode changes.